File: 0234-palindrome-linked-list/0234-palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
const isPalindrome = function(head) {
    if (!head || !head.next) return true;
    const reverseRecursively = (node) => {
        if (node === null || node.next === null) {
            return node;
        }
        const newHead = reverseRecursively(node.next);
        node.next.next = node;
        node.next = null;
        return newHead;
    };

    let slow = head;
    let fast = head;
    
    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
    }

    let reversedHalf = reverseRecursively(slow);

    let firstHalf = head;
    let secondHalf = reversedHalf;

    while (secondHalf) {
        if (firstHalf.val !== secondHalf.val) {
            return false;
        }
        firstHalf = firstHalf.next;
        secondHalf = secondHalf.next;
    }

    return true;
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverse(self,node):
        if node is None or node.next is None:
            return node

        newhead = self.reverse(node.next)
        node.next.next = node
        node.next = None
        return newhead

    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        isOdd = False

        if fast is not None:
            isOdd = True
        
        if isOdd:
            slow = slow.next

        first = head
        second = self.reverse(slow)

        while second:
            if first.val != second.val:
                return False

            first = first.next
            second = second.next
        
        return True